packages/source-iotsitewise/src/time-series-data/client/batchGetHistoricalPropertyDataPoints.ts

```diff
@@ -0,0 +1,157 @@
+import {
+  BatchGetAssetPropertyValueHistoryCommand,
+  BatchGetAssetPropertyValueHistoryErrorEntry,
+  BatchGetAssetPropertyValueHistorySuccessEntry,
+  IoTSiteWiseClient,
+  TimeOrdering,
+} from '@aws-sdk/client-iotsitewise';
+import { toDataPoint } from '../util/toDataPoint';
+import { dataStreamFromSiteWise } from '../dataStreamFromSiteWise';
+import { OnSuccessCallback, ErrorCallback, RequestInformationAndRange } from '@iot-app-kit/core';
+import { toSiteWiseAssetProperty } from '../util/dataStreamId';
+import { isDefined } from '../../common/predicates';
+import { HistoricalPropertyParams } from './client';
+import { createEntryBatches } from './batch';
+
+type BatchHistoricalEntry = {
+  requestInformation: RequestInformationAndRange;
+  maxResults?: number;
+  onError: ErrorCallback;
+  onSuccess: OnSuccessCallback;
+  requestStart: Date;
+  requestEnd: Date;
+};
+
+type BatchEntryCallbackCache = {
+  [key: string]: {
+    onError: (entry: BatchGetAssetPropertyValueHistoryErrorEntry) => void;
+    onSuccess: (entry: BatchGetAssetPropertyValueHistorySuccessEntry) => void;
+  };
+};
+
+const batchGetHistoricalPropertyDataPointsForProperty = async ({
+  client,
+  entries,
+  nextToken: prevToken,
+}: {
+  client: IoTSiteWiseClient;
+  entries: BatchHistoricalEntry[];
+  nextToken?: string;
+}) => {
+  // callback cache makes it convenient to capture request data in a closure.
+  // the cache exposes methods that only require batch response entry as an argument.
+  const callbackCache: BatchEntryCallbackCache = {};
+
+  return Promise.all(
+    createEntryBatches<BatchHistoricalEntry>(entries)
+      .filter((batch) => batch.length > 0) // filter out empty batches
+      .map(([batch, maxResults], requestIndex) =>
+        client
+          .send(
+            new BatchGetAssetPropertyValueHistoryCommand({
+              entries: batch.map((entry, entryIndex) => {
+                const { requestInformation, onError, onSuccess, requestStart, requestEnd } = entry;
+                const { id } = requestInformation;
+
+                // use 2D array indices as entryIDs to guarantee uniqueness
+                // entryId is used to map batch entries with the appropriate callback
+                const entryId = String(`${requestIndex}-${entryIndex}`);
+
+                // save request entry data in functional closure.
+                callbackCache[entryId] = {
+                  onError: ({ errorMessage: msg = 'batch error', errorCode: status }) => {
+                    onError({
+                      id,
+                      resolution: 0,
```

Comment:
is it correct to always report the raw data as what's failing?

Reply:
yeah, this mirrors the existing behaviour - https://github.com/awslabs/iot-app-kit/blob/main/packages/source-iotsitewise/src/time-series-data/client/getHistoricalPropertyDataPoints.ts#L78.

this particular path only handles raw data, there will be a similar method for handling aggregated data coming in a future PR